include/pybind11/detail/internals.h

```diff
@@ -419,11 +440,22 @@ PYBIND11_NOINLINE internals &get_internals() {
     } gil;
     error_scope err_scope;
 
-    PYBIND11_STR_TYPE id(PYBIND11_INTERNALS_ID);
-    auto builtins = handle(PyEval_GetBuiltins());
-    if (builtins.contains(id) && isinstance<capsule>(builtins[id])) {
-        internals_pp = static_cast<internals **>(capsule(builtins[id]));
+    constexpr const char *id_cstr = PYBIND11_INTERNALS_ID;
+    str id(id_cstr);
+
+    dict state_dict = get_internals_state_dict();
+
+    if (state_dict.contains(id_cstr)) {
+        object o = state_dict[id];
+        // May fail if 'capsule_obj' is not a capsule, or if it has a different
+        // name. We clear the error status below in that case
+        internals_pp = static_cast<internals **>(PyCapsule_GetPointer(o.ptr(), id_cstr));
+        if (!internals_pp) {
+            PyErr_Clear();
```

Comment:
This PR (with the #4307 tweak) is now baked into PR #4329:

c2e6b38c9376cae59fb803793a6b49f8f8a3351f

Copying the commit message here (I think the link will stop working in case I have to rebase):
________
    Modified version of PR #4293 by @wjakob

    Modifications are:

    * Backward compatibility (no ABI break), as originally under PR #4307.
    * Naming: `get_python_state_dict()`, `has_pybind11_internals_capsule()`
    * Report error retrieving `internals**` from capsule instead of clearing it.

    Locally tested with ASAN, MSAN, TSAN, UBSAN (Google-internal toolchain).
________

My commit changes the code here, to report the error rather than suppressing it.

What is the rationale for suppressing it?